### PR TITLE
SmbShare module new compression parameters

### PR DESCRIPTION
**Pull request by robinharwood**
The [SmbShare Module | Microsoft Docs](https://docs.microsoft.com/en-us/powershell/module/smbshare/?view=windowsserver2022-ps) for Windows Server needs to be updated with new parameters for the Reset-SmbClientConfiguration, Reset-SmbServerConfiguration, Set-SmbClientConfiguration, and Set-SmbServerConfiguration commands.

More information can be found in the blog article: [SMB compression behavior & settings changes - Microsoft Tech Community](https://techcommunity.microsoft.com/t5/storage-at-microsoft/smb-compression-behavior-amp-settings-changes/ba-p/3610763)

### CODE REVIEW

**Comment by robinharwood** [comment]
Hi @nedpyle - I can't assigned you for review in this repo but hopefully tagging you here will help. This is the PR to update the SMB compression parameters. Please let me know if you have a good example to add and let me know your thoughts on the parameter text. Of course, I'm happy for us to expand these as discussed.

Thanks,
Robin

File: docset/winserver2022-ps/smbshare/Set-SmbClientConfiguration.md [hunk 1]
```diff
@@ -229,6 +264,27 @@ Accept pipeline input: False
 Accept wildcard characters: False
 ```
 
+### -EnableCompressibilitySampling
+Controls the sampling behavior. SMB by default always attempts to compress the entire file when a
+client or server requests it, without using compression sampling. With EnableCompressibilitySampling
+set to '$true' SMB uses of an algorithm where it attempted to compress the first 524,288,000 bytes
```
Comment (nedpyle) [inline docset/winserver2022-ps/smbshare/Set-SmbClientConfiguration.md:270]: default algorithm. Those values of 500MiB and 100MiB are what are controlled by CompressibiltySamplingSize and CompressibleThreshold above - if not populated, these are the values used (maybe that should be in their example?).

 And despite their ranges being UInt64, the _actual_ limit for the compression feature is 9,007,199,254,740,992 (9PiB). That's worth nothing on those two params help.

**Review by nedpyle — COMMENTED** [review]
All set.

File: docset/winserver2022-ps/smbshare/New-SmbShare.md [hunk 2]
```diff
@@ -314,9 +328,9 @@ Accept wildcard characters: False
 ```
 
 ### -Name
-Specifies a name for the SMB share.
-The name may be composed of any valid file name characters, but must be less than 80 characters long.
-The names `pipe` and `mailslot` are reserved for use by the computer.
+Specifies a name for the SMB share. The name may be composed of any valid file name characters, but
+must be less than 80 characters long. The names `pipe` and `mailslot` are reserved for use by the
```
[the remark this answers is not in the document]
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/New-SmbShare.md]: I've actually reworded a good portion of this in the next commit. Please let me know if you are happy with the updates.

File: docset/winserver2022-ps/smbshare/New-SmbShare.md [hunk 3]
```diff
@@ -17,56 +17,68 @@ Creates an SMB share.
...
+The `New-SmbShare` cmdlet exposes a file system folder to remote clients as a Server Message Block
+(SMB) share.
 
-To delete a share that was created by this cmdlet, use the Remove-SmbShare cmdlet.
+To delete a share that was created by this cmdlet, use the `Remove-SmbShare` cmdlet.
 
 ## EXAMPLES
 
 ### Example 1: Create an SMB share
+```powershell
```
[the remark this answers is not in the document]
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/New-SmbShare.md:40]: As discussed, whilst I do agree this is part of the repo guidelines to improve readability when rendered by the Learn engine and when viewed from a PowerShell session.

File: docset/winserver2022-ps/smbshare/New-SmbShare.md [hunk 4]
```diff
@@ -17,56 +17,68 @@ Creates an SMB share.
...
 This command creates an encrypted SMB share.
 
 ### Example 3: Create an SMB share with Multiple Permissions
-```
-PS C:\>New-SmbShare -Name "VMSFiles" -Path "C:\ClusterStorage\Volume1\VMFiles" -ChangeAccess "Users" -FullAccess "Administrators"
-
-Name                          ScopeName                     Path                          Description
-----                          ---------                     ----                          -----------
-VMSFiles                      Contoso-SO                    C:\ClusterStorage\Volume1\...
+```powershell
```
[the remark this answers is not in the document]
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/New-SmbShare.md:65]: As discussed, whilst I do agree this is part of the repo guidelines to improve readability when rendered by the Learn engine and when viewed from a PowerShell session.

File: docset/winserver2022-ps/smbshare/New-SmbShare.md [hunk 5]
```diff
@@ -277,9 +288,11 @@ Accept wildcard characters: False
 ```
 
 ### -FullAccess
-Specifies which accounts are granted full permission to access the share.
-Use a comma-separated list to specify multiple accounts.
-An account may not be specified more than once in the *FullAccess*, *ChangeAccess*, or *ReadAccess* parameter lists, but may be specified once in the *FullAccess*, *ChangeAccess*, or *ReadAccess* parameter list and once in the *NoAccess* parameter list.
+Specifies which accounts are granted full permission to access the share. Use a comma-separated list
+to specify multiple accounts. An account may not be specified more than once in the **FullAccess**,
+**ChangeAccess**, or **ReadAccess** parameter lists, but may be specified once in the
+**FullAccess**, **ChangeAccess**, or **ReadAccess** parameter list and once in the **NoAccess**
```
[the remark this answers is not in the document]
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/New-SmbShare.md:311]: This is part of the repo guidelines. Line feeds help with readability especially within the GitHub UI.

File: docset/winserver2022-ps/smbshare/Set-SmbClientConfiguration.md (hunk 1 above)
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/Set-SmbClientConfiguration.md]: I've updated this whole parameter; I've also referenced the actual compression limit in the other two parameters. Please let me know if you are happy with the updates.

**Review by nedpyle — APPROVED** [review]
So many contractions! :D

**Comment by robinharwood** [comment]
Hi @michaeltlombardi / @sdwheeler,

I believe this PR is release ready if you would be kind enough to review and if appropriate sign-off/merge 😊

The main update was to add in new compression parameters, I have also addressed some quality issues at the same time.

Please let me know if you have any questions or if I need to apply any other updates.

Thanks Robin

File: docset/winserver2022-ps/smbshare/Set-SmbClientConfiguration.md [hunk 6]
```diff
@@ -229,6 +272,33 @@ Accept pipeline input: False
...
+compression sampling algorithm where it attempts to compress the file based on the values
+configured in the **CompressibiltySamplingSize** and **CompressibleThreshold** parameters.
+
+With **EnableCompressibilitySampling** set to `$true` and if you don't use the
+**CompressibiltySamplingSize** and **CompressibleThreshold** parameters, the algorithm will by
+default attempt to compress the first 524,288,000 bytes (500 MiB) of a file during transfer. The
+algorithm will try to track that at least 104,857,600 bytes (100 MiB) compresses within that 500 MiB
+range. If fewer than 100 MiB was compressible, SMB compression stopped trying to compress the rest
+of the file. If at least 100 MiB compressed, SMB compression attempted to compress the rest of the
+file.
```
Comment (michaeltlombardi) [inline docset/winserver2022-ps/smbshare/Set-SmbClientConfiguration.md:288]:
```suggestion
When you specify **EnableCompressibilitySampling** as `$true` and don't specify either the
**CompressibiltySamplingSize** or **CompressibleThreshold** parameters, the algorithm attempts to
compress the first `524,288,000` bytes (500 MiB) of a file during transfer. The algorithm tries to
track that at least `104,857,600` bytes (100 MiB) compresses within that 500 MiB range. If fewer
than 100 MiB was compressible, SMB compression stops trying to compress the rest of the file. If at
least 100 MiB compressed, SMB compression attempts to compress the rest of the file.
```

File: docset/winserver2022-ps/smbshare/Set-SmbServerConfiguration.md [hunk 7]
```diff
@@ -46,10 +47,15 @@ For more information on SMB server and protocol specifications, see
...
-> The **EncryptionCiphers** parameter is available beginning with 2022-06 Cumulative Update for
+> - The **EncryptionCiphers** parameter is available beginning with 2022-06 Cumulative Update for
 > Microsoft server operating system version 21H2 for x64-based Systems
 > ([KB5014665](https://support.microsoft.com/help/5014665)), and Cumulative Update for Windows 11,
 > version 22H2 ([KB5014668](https://support.microsoft.com/help/5014668)).
+>
+> - The **DisableCompression** and **RequestCompression** parameters are available beginning with
+> 2022-08 Cumulative Update for Microsoft server operating system version 21H2 for x64-based Systems
+> ([KB5016693](https://support.microsoft.com/help/5016693)), and Cumulative Update for Windows 11,
+> version 22H2 ([KB5016691](https://support.microsoft.com/help/5016691)).
```
Comment (michaeltlombardi) [inline docset/winserver2022-ps/smbshare/Set-SmbServerConfiguration.md:58]:
```suggestion
> - The **DisableCompression** and **RequestCompression** parameters are available beginning with
>   2022-08 Cumulative Update for Microsoft server operating system version 21H2 for x64-based
>   Systems ([KB5016693](https://support.microsoft.com/help/5016693)), and Cumulative Update for
>   Windows 11, version 22H2 ([KB5016691](https://support.microsoft.com/help/5016691)).
```

File: docset/winserver2022-ps/smbshare/Set-SmbServerConfiguration.md [hunk 8]
```diff
@@ -277,6 +283,23 @@ Accept pipeline input: False
 Accept wildcard characters: False
 ```
 
+### -DisableCompression
+
+Indicated that the SMB server should never compress files even if client or application requested
```
Comment (michaeltlombardi) [inline docset/winserver2022-ps/smbshare/Set-SmbServerConfiguration.md:288]:
```suggestion
Indicates that the SMB server should never compress files even if client or application requested
```

File: docset/winserver2022-ps/smbshare/Reset-SmbClientConfiguration.md [hunk 9]
```diff
@@ -103,6 +108,22 @@ Accept pipeline input: False
 Accept wildcard characters: False
 ```
 
+### -CompressibilitySampling
+
+Resets the compression sampling behavior.
```
Comment (michaeltlombardi) [inline docset/winserver2022-ps/smbshare/Reset-SmbClientConfiguration.md:113]: Should this match the other parameters?
```suggestion
Resets the compression sampling behavior to its default value.
```

File: docset/winserver2022-ps/smbshare/New-SmbShare.md [hunk 10]
```diff
@@ -277,9 +304,12 @@ Accept wildcard characters: False
...
 ### -FullAccess
-Specifies which accounts are granted full permission to access the share.
-Use a comma-separated list to specify multiple accounts.
-An account may not be specified more than once in the *FullAccess*, *ChangeAccess*, or *ReadAccess* parameter lists, but may be specified once in the *FullAccess*, *ChangeAccess*, or *ReadAccess* parameter list and once in the *NoAccess* parameter list.
+
+Specifies which accounts are granted full permission to access the share. Use a comma-separated list
+to specify multiple accounts. An account may not be specified more than once in the **FullAccess**,
+**ChangeAccess**, or **ReadAccess** parameter lists, but may be specified once in the
+**FullAccess**, **ChangeAccess**, or **ReadAccess** parameter list and once in the **NoAccess**
+parameter list.
```
Comment (michaeltlombardi) [inline docset/winserver2022-ps/smbshare/New-SmbShare.md:312]: Suggest striking the second sentence.
```suggestion
Specifies which accounts are granted full permission to access the share. An account may not be
specified more than once in the **FullAccess**, **ChangeAccess**, or **ReadAccess** parameter
lists, but may be specified once in the **FullAccess**, **ChangeAccess**, or **ReadAccess**
parameter list and once in the **NoAccess** parameter list.
```

File: docset/winserver2022-ps/smbshare/Set-SmbServerConfiguration.md [hunk 11]
```diff
@@ -715,7 +738,7 @@ Accept wildcard characters: False
 
 ### -RejectUnencryptedAccess
 
-Indicates whether the client that does not support encryption is denied access if it attempts to
+Indicates whether the client that doesn't support encryption is denied access if it attempts to
```
Comment (michaeltlombardi) [inline docset/winserver2022-ps/smbshare/Set-SmbServerConfiguration.md:741]:
```suggestion
Indicates whether a client that doesn't support encryption is denied access if it attempts to
```

File: docset/winserver2022-ps/smbshare/Set-SmbServerConfiguration.md [hunk 12]
```diff
@@ -730,6 +753,23 @@ Accept pipeline input: False
 Accept wildcard characters: False
 ```
 
+### -RequestCompression
+
+Indicates if SMB server should always request compression even if client or application didn't
+specify it.
```
Comment (michaeltlombardi) [inline docset/winserver2022-ps/smbshare/Set-SmbServerConfiguration.md:759]:
```suggestion
```suggestion
Indicates whether the SMB server should always request compression even if client or application
didn't specify it.
```
```

**Review by michaeltlombardi — CHANGES_REQUESTED** [review]
Some minor formatting/consistency notes and a few wording suggestions, but otherwise this looks great!

File: docset/winserver2022-ps/smbshare/New-SmbShare.md [hunk 13]
```diff
@@ -219,10 +244,10 @@ Accept wildcard characters: False
...
 
 ### -Description
-Specifies an optional description of the SMB share.
-A description of the share is displayed by running the Get-SmbShare cmdlet.
-The description may not contain more than 256 characters.
-The default value no description, or an empty description.
+
+Specifies an optional description of the SMB share. A description of the share is displayed by
+running the `Get-SmbShare` cmdlet. The description may not contain more than 256 characters. The
+default value no description, or an empty description.
```
Comment (JohanFreelancer9) [inline docset/winserver2022-ps/smbshare/New-SmbShare.md:250]:
```suggestion
default value is no description or an empty description.
```

File: docset/winserver2022-ps/smbshare/New-SmbShare.md [hunk 14]
```diff
@@ -314,9 +345,11 @@ Accept wildcard characters: False
 ```
 
 ### -Name
-Specifies a name for the SMB share.
-The name may be composed of any valid file name characters, but must be less than 80 characters long.
-The names `pipe` and `mailslot` are reserved for use by the computer.
+
+Specifies a name for the SMB share. The names `pipe` and `mailslot` are reserved for use by the
+computer. Share names can be up to a maximum of 80 characters long. The SMB share name can use any
+character allowed by Windows for files and directories, to learn more about naming conventions for
```
Comment (JohanFreelancer9) [inline docset/winserver2022-ps/smbshare/New-SmbShare.md:351]:
```suggestion
character allowed by Windows for files and directories. To learn more about naming conventions for
```

File: docset/winserver2022-ps/smbshare/New-SmbShare.md [hunk 15]
```diff
@@ -427,9 +466,12 @@ Accept wildcard characters: False
 ```
 
 ### -ThrottleLimit
-Specifies the maximum number of concurrent operations that can be established to run the cmdlet.
-If this parameter is omitted or a value of `0` is entered, then Windows PowerShell® calculates an optimum throttle limit for the cmdlet based on the number of CIM cmdlets that are running on the computer.
-The throttle limit applies only to the current cmdlet, not to the session or to the computer.
+
+Specifies the maximum number of concurrent operations that can be established to run the cmdlet. If
+this parameter is omitted or a value of `0` is entered, then Windows PowerShell® calculates an
```
Comment (JohanFreelancer9) [inline docset/winserver2022-ps/smbshare/New-SmbShare.md:471]:
```suggestion
this parameter is omitted or a value of `0` is entered, then Windows PowerShell calculates an
```

File: docset/winserver2022-ps/smbshare/Set-SmbClientConfiguration.md [hunk 16]
```diff
@@ -502,6 +572,23 @@ Accept pipeline input: False
 Accept wildcard characters: False
 ```
 
+### -RequestCompression
+
+Indicates if SMB client should always request compression even if server or application didn't
```
Comment (JohanFreelancer9) [inline docset/winserver2022-ps/smbshare/Set-SmbClientConfiguration.md:577]:
```suggestion
Indicates if SMB client should always request compression even if the server or application didn't
```

**Review by JohanFreelancer9 — CHANGES_REQUESTED** [review]
Copy review - @robinharwood, kindly commit the suggested changes. Many thanks.

**Review by JohanFreelancer9 — APPROVED** [review]

File: docset/winserver2022-ps/smbshare/New-SmbShare.md [hunk 17]
```diff
@@ -139,8 +158,9 @@ Accept wildcard characters: False
 ```
 
 ### -ChangeAccess
-Specifies which users are granted modify permission to access the share.
-Multiple users can be specified by using a comma-separated list.
+
+Specifies which users are granted modify permission to access the share. Multiple users can be
+specified by using a comma-separated list.
```
[the remark this answers is not in the document]
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/New-SmbShare.md:163]: It is kind of like that. I've updated the text as below:

> Specifies which accounts are granted full permission to access the share. Use a comma-separated list
to specify multiple accounts. Each entry in the comma-separated list must be contained within single
or double quotes, for example `'CONTOSO\Finance Users','CONTOSO\HR Users'`.

File: docset/winserver2022-ps/smbshare/New-SmbShare.md (hunk 10 above)
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/New-SmbShare.md:312]: It is kind of like that. I've updated the text as below:

> Specifies which accounts are granted full permission to access the share. Use a comma-separated list
to specify multiple accounts. Each entry in the comma-separated list must be contained within single
or double quotes, for example `'CONTOSO\Finance Users','CONTOSO\HR Users'`.
> 
> An account may not be specified more than once in the **FullAccess**, **ChangeAccess**, or
**ReadAccess** parameter lists, but may be specified once in the **FullAccess**, **ChangeAccess**,
or **ReadAccess** parameter list and once in the **NoAccess** parameter list.

File: docset/winserver2022-ps/smbshare/New-SmbShare.md [hunk 18]
```diff
@@ -252,14 +278,15 @@ Accept wildcard characters: False
...
-- Unrestricted.
-SMB displays files and folders to a user even when the user does not have permission to access the items.
+Specifies which files and folders in the SMB share are visible to users. The acceptable values for
+this parameter are:
+
+- AccessBased. SMB doesn't display the files and folders for a share to a user unless that user has
+  rights to access the files and folders. By default, access-based enumeration is disabled for new
+  SMB shares.
+- Unrestricted. SMB displays files and folders to a user even when the user doesn't have permission
+  to access the items.
```
[the remark this answers is not in the document]
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/New-SmbShare.md]: Updated as suggested

File: docset/winserver2022-ps/smbshare/New-SmbShare.md [hunk 19]
```diff
@@ -331,6 +364,7 @@ Accept wildcard characters: False
 ```
 
 ### -NoAccess
+
 Specifies which accounts are denied access to the SMB share.
 Multiple accounts can be specified by supplying a comma-separated list.
```
[the remark this answers is not in the document]
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/New-SmbShare.md:369]: Kind of, updated to:

>Specifies which accounts are denied access to the SMB share. Multiple accounts can be specified by
supplying a comma-separated list. Each entry in the comma-separated list must be contained within
single or double quotes, for example `'CONTOSO\Finance Users','CONTOSO\HR Users'`.

File: docset/winserver2022-ps/smbshare/Set-SmbServerConfiguration.md (hunk 8 above)
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/Set-SmbServerConfiguration.md]: Updated as suggested.

File: docset/winserver2022-ps/smbshare/Set-SmbServerConfiguration.md (hunk 11 above)
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/Set-SmbServerConfiguration.md]: Updated as sugggested.

File: docset/winserver2022-ps/smbshare/Set-SmbServerConfiguration.md (hunk 12 above)
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/Set-SmbServerConfiguration.md]: Updated as suggested.

File: docset/winserver2022-ps/smbshare/Set-SmbClientConfiguration.md (hunk 16 above)
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/Set-SmbClientConfiguration.md]: Updated as suggested.

**Comment by robinharwood** [comment]
Thanks @michaeltlombardi and @JohanFreelancer9 - Updates applied, I believe this is now ready to merge. #sign-off

File: docset/winserver2022-ps/smbshare/New-SmbShare.md [hunk 20]
```diff
@@ -410,8 +448,9 @@ Accept wildcard characters: False
 ```
 
 ### -Temporary
+
 Specifies the lifetime of the new SMB share.
-A temporary share does not persist beyond the next restart of the computer.
+A temporary share doesn't persist beyond the next restart of the computer.
 By default, new SMB shares are persistent, and non-temporary.
```
[the remark this answers is not in the document]
Comment (michaeltlombardi) [reply docset/winserver2022-ps/smbshare/New-SmbShare.md:459]: @robinharwood I think this should be addressed for clarity.

[the remark this answers is not in the document]
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/New-SmbShare.md:459]: Sorry @michaeltlombardi, I missed a few of these when I did the rebase and force-push. I've updated this to say:

`Specifies the new SMB share is temporary and will not persist beyond the next restart of the
computer. By default, new SMB shares aren't temporary.`

File: docset/winserver2022-ps/smbshare/New-SmbShare.md [hunk 21]
```diff
@@ -364,6 +399,7 @@ Accept wildcard characters: False
 ```
 
 ### -ReadAccess
+
 Specifies which users are granted read permission to access the share.
 Multiple users can be specified by supplying a comma-separated list.
```
[the remark this answers is not in the document]
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/New-SmbShare.md]: Updated text to clarify the use of comma-separated items.

File: docset/winserver2022-ps/smbshare/New-SmbShare.md [hunk 22]
```diff
@@ -459,10 +502,11 @@ Accept wildcard characters: False
 ```
 
 ### -WhatIf
+
 Shows what would happen if the cmdlet runs.
-The cmdlet is not run.
+The cmdlet isn't run.
 
-NOTE: The WhatIf switch does not work with this cmdlet.
+NOTE: The WhatIf switch doesn't work with this cmdlet.
```
[the remark this answers is not in the document]
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/New-SmbShare.md:514]: I've updated the parameter to be in bold. I didn't think you could put alert boxes in the parameter sections (I will test will a follow up commit)

[the remark this answers is not in the document]
Comment (robinharwood) [reply docset/winserver2022-ps/smbshare/New-SmbShare.md]: Added the alert box and verified the build completed.

**Comment by robinharwood** [comment]
Hi @michaeltlombardi - I think all comments are now addressed. Thanks, Robin

**Review by michaeltlombardi — APPROVED** [review]

**Comment by robinharwood** [comment]
Thanks for approving @michaeltlombardi - are you able to do the #sign-off and merge? I don't have the permissions for this repo.

**Comment by michaeltlombardi** [comment]
@robinharwood sorry, thought I hit the button before closing the tab! 😣

**Pull request merged**